File: linked_lists/705. 设计哈希集合.go
package linked_lists

import "container/list"

const base = 769

func hash(key int) int {
	return key % base
}

type MyHashSet struct {
	data []list.List
}

func NewMyHashSet() MyHashSet {
	return MyHashSet{make([]list.List, base)}
}
func (s *MyHashSet) Add(key int) {
	if s.Contains(key) {
		return
	}
	h := hash(key)
	s.data[h].PushBack(key)
}
func (s *MyHashSet) Remove(key int) {
	h := hash(key)
	for e := s.data[h].Front(); e != nil; e = e.Next() {
		if e.Value == key {
			s.data[h].Remove(e)
			return
		}
	}
}
func (s *MyHashSet) Contains(key int) bool {
	h := hash(key)
	for e := s.data[h].Front(); e != nil; e = e.Next() {
		if e.Value == key {
			return true
		}
	}
	return false
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(key);
 * obj.Remove(key);
 * param_3 := obj.Contains(key);
 */

File: linked_lists/705. 设计哈希集合.cc
//go:build ignore

// 方法一：链地址法
class MyHashSet {
private:
    static const int base = 769;
    vector<list<int>> data;
    int hash(int key) {
        return key % base;
    }
public:
    MyHashSet(): data(base) {}
    void add(int key) {
        if (contains(key)) return;
        int h = hash(key);
        data[h].push_back(key);
    }
    void remove(int key) {
        int h = hash(key);
        for (auto it = data[h].begin(); it != data[h].end(); ++it) {
            if ((*it) == key) {
                data[h].erase(it);
                return;
            }
        }
    }
    bool contains(int key) {
        int h = hash(key);
        for (auto it = data[h].begin(); it != data[h].end(); ++it) {
            if ((*it) == key) {
                return true;
            }
        }
        return false;
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */